### PR TITLE
Binding

### PR DESCRIPTION
**Pull request by KarlLeell**

### CODE REVIEW

File: tmp/ambient_occlusion.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+npe_begin_code()
+
+  EigenDense<npe_Scalar_p> s;
+  igl::ambient_occlusion(aabb, v, f, p, n, num_samples, s);
+  return npe::move(s);
+
+npe_end_code()
+
+
```
Comment (teseoch) [inline tmp/ambient_occlusion.cpp]: let's keep only the first one for the moment and leve a todo note on top

Comment (KarlLeell) [reply tmp/ambient_occlusion.cpp]: you mean try to fix the first one without those messy types, comment out the rest two and move it to src?

File: tmp/bounding_box.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+const char* ds_bounding_box_pad = R"igl_Qu8mg5v7(
+See bounding_box for the documentation.
+)igl_Qu8mg5v7";
+
+npe_function(bounding_box)
+npe_doc(ds_bounding_box)
+
+npe_arg(v, dense_float, dense_double)
+npe_arg(pad, typename DerivedV::Scalar)
```
Comment (fwilliams) [inline tmp/bounding_box.cpp]: replace with `double`

File: tmp/ambient_occlusion.cpp [hunk 3]
```diff
@@ -1,144 +0,0 @@
-// TODO: Fix the npe_arg for igl::AABB and std::function
```
Comment (fwilliams) [inline tmp/ambient_occlusion.cpp]: Are we keeping any version of this function or just killing all of it? The VFPN version (first function) is probably useful and easy to bind.

File: tmp/biharmonic_coordinates.cpp [hunk 4]
```diff
@@ -81,23 +82,25 @@ npe_doc(ds_biharmonic_coordinates)
 
 npe_arg(v, dense_float, dense_double)
 npe_arg(t, dense_float, dense_double)
-npe_arg(s, std::vector<std::vector<SType> > &)
+npe_arg(s, std::vector<std::vector<int> > &)
```
Comment (fwilliams) [inline tmp/biharmonic_coordinates.cpp]: Let's add a unit test to make sure this gets properly bound with pybind (try with ints and longs just to check)

File: tmp/biharmonic_coordinates.cpp [hunk 5]
```diff
@@ -116,18 +119,18 @@ Examples
...
 
-npe_function(biharmonic_coordinates)
-npe_doc(ds_biharmonic_coordinates)
+npe_function(bitriharmonic_coordinates)
+npe_doc(ds_bitriharmonic_coordinates)
 
 npe_arg(v, dense_float, dense_double)
 npe_arg(t, dense_float, dense_double)
-npe_arg(s, std::vector<std::vector<SType> > &)
+npe_arg(s, std::vector<std::vector<int> > &)
```
Comment (fwilliams) [inline tmp/biharmonic_coordinates.cpp]: Same comment as above. Let's make sure this works with a python list of lists.

File: tmp/ambient_occlusion.cpp (hunk 3 above)
Comment (KarlLeell) [reply tmp/ambient_occlusion.cpp]: Resolved this in the commit this morning. Now there is only the VFPN function, and  I have removed the file from tmp directory.

File: tmp/biharmonic_coordinates.cpp (hunk 4 above)
Comment (KarlLeell) [reply tmp/biharmonic_coordinates.cpp]: This part I'm still debugging, it fails to compile, so it is in the tmp directory which will not be compiled.

File: src/circulation.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+
+npe_begin_code()
+
+  // TODO: remove __copy
+  Eigen::VectorXi emap_copy = emap.template cast<int>();
+  Eigen::MatrixXi ef_copy = ef.template cast<int>();
+  Eigen::MatrixXi ei_copy = ei.template cast<int>();
+  igl::circulation(e, ccw, emap_copy, ef_copy, ei_copy);
+  return ;
```
Comment (teseoch) [inline src/circulation.cpp]: `return igl::circulation(e, ccw, emap_copy, ef_copy, ei_copy);`

and drop other

File: src/boundary_conditions.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  Eigen::MatrixXd v_copy = v.template cast<double>();
+  Eigen::MatrixXi ele_copy = ele.template cast<int>();
+  Eigen::MatrixXd c_copy = c.template cast<double>();
+  Eigen::VectorXi p_copy = p.template cast<int>();
+  Eigen::MatrixXi be_copy = be.template cast<int>();
+  Eigen::MatrixXi ce_copy = ce.template cast<int>();
+  Eigen::VectorXi b;
+  Eigen::MatrixXd bc;
+  igl::boundary_conditions(v_copy, ele_copy, c_copy, p_copy, be_copy, ce_copy, b, bc);
+  return std::make_tuple(npe::move(p_copy), npe::move(v_copy));
```
Comment (teseoch) [inline src/boundary_conditions.cpp]: return B and BC

File: src/boundary_conditions.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  //TODO: remove __copy
+  Eigen::MatrixXd v_copy = v.template cast<double>();
+  Eigen::MatrixXi ele_copy = ele.template cast<int>();
+  Eigen::MatrixXd c_copy = c.template cast<double>();
+  Eigen::VectorXi p_copy = p.template cast<int>();
+  Eigen::MatrixXi be_copy = be.template cast<int>();
+  Eigen::MatrixXi ce_copy = ce.template cast<int>();
+  Eigen::VectorXi b;
+  Eigen::MatrixXd bc;
+  igl::boundary_conditions(v_copy, ele_copy, c_copy, p_copy, be_copy, ce_copy, b, bc);
```
Comment (teseoch) [inline src/boundary_conditions.cpp]: `bool ok = igl::boundary_conditions(v_copy, ele_copy, c_copy, p_copy, be_copy, ce_copy, b, bc);`
return B, Bc, and ok

**Pull request merged**